### PR TITLE
Adding segregation of resultType for CloudError

### PR DESCRIPTION
**Pull request by sankur-codes**
### Which issue this PR addresses:

Fixes [ARO-5112](https://issues.redhat.com/browse/ARO-5112)

### What this PR does / why we need it:

As of now, all CloudErrors are termed as UserError.  This PR will check upon the statusCode of the error and provide resultType accordingly. 

### Test plan for issue:

Done : Manually Tested mapping of CloudError `StatusCode` with `ResultType`.
Also, included test coverage for the functions
### Is there any documentation that needs to be updated for this PR?

No

### CODE REVIEW

File: pkg/frontend/frontend.go [hunk 1]
```diff
@@ -475,6 +475,14 @@ func frontendOperationResultLog(log *logrus.Entry, method string, err error) {
 
 	switch err := err.(type) {
 	case *api.CloudError:
+		if int(err.StatusCode) < 300 && int(err.StatusCode) >= 200 {
```
Comment (mociarain) [inline pkg/frontend/frontend.go:478]: Can we really have a "successful" CloudError?

**Review by mociarain — COMMENTED** [review]
Looks good but I have a small suggestion. Would we consider putting this into the `CloudError` itself?

We could have a:
```
func (err *CloudError) GetResultType() ResultType {...}
```

This way the logic of mapping stausCode to ResultType lives in one place i.e. the thing that holds the statusCode.

File: pkg/frontend/frontend.go (hunk 1 above)
Comment (sankur-codes) [reply pkg/frontend/frontend.go:478]: Considering the condition, there could not be a "real" error but a CloudError that has a statusCode between 200 & 300. 
Eg : https://github.com/Azure/ARO-RP/blob/master/pkg/frontend/admin_hive_clusterdeployment_get.go#L45

Comment (mociarain) [reply pkg/frontend/frontend.go:478]: An error with a 204 seems like a bug to me and IMO we should make that impossible. Definitely not for this PR though. 

Also after a quick grep I think you might have the only instance of the a CloudError with a non 4XX/5XX status code in the whole RP so maybe the change I'm implying wouldn't be so painful. If you have sometime maybe we could pair on this as a subsequent task? I don't think I've met any of the APAC folks yet.

**Comment by sankur-codes** [comment]
> Looks good but I have a small suggestion. Would we consider putting this into the `CloudError` itself?
> 
> We could have a:
> 
> ```
> func (err *CloudError) GetResultType() ResultType {...}
> ```
> 
> This way the logic of mapping stausCode to ResultType lives in one place i.e. the thing that holds the statusCode.

**Comment by sankur-codes** [comment]
> Looks good but I have a small suggestion. Would we consider putting this into the `CloudError` itself?
> 
> We could have a:
> 
> ```
> func (err *CloudError) GetResultType() ResultType {...}
> ```
> 
> This way the logic of mapping stausCode to ResultType lives in one place i.e. the thing that holds the statusCode.

Putting a high level view on the suggestion: To include `ResultType` as part of or use it in ClousError, `github.com/Azure/ARO-RP/pkg/util/log` will have to be imported in `pkg/api/error.go` which might create a circular dependency leading to import cycle. 
Another option would be to have `ResultType` completely under CloudError and defined in `pkg/api/error.go` itself but that might not be totally accurate as `ResultType` is meant to be part of **logging** & thus placed in `pkg/util/log`.
I might be missing on something, do help me with further suggestions.

**Comment by mociarain** [comment]
> Putting a high level view on the suggestion: To include `ResultType` as part of or use it in ClousError, `github.com/Azure/ARO-RP/pkg/util/log` will have to be imported in `pkg/api/error.go` which might create a circular dependency leading to import cycle. Another option would be to have `ResultType` completely under CloudError and defined in `pkg/api/error.go` itself but that might not be totally accurate as `ResultType` is meant to be part of **logging** & thus placed in `pkg/util/log`. I might be missing on something, do help me with further suggestions.

No stress. We could put something in `log` that maps `statusCode int` to `ResultType` instead. IMO so long as this logic is in one place I think that's a win. We can do this as part of the pairing on Friday actually (or as a separate PR) since it's gonna branch out beyond this PR

**Review by mociarain — APPROVED** [review]

File: pkg/backend/openshiftcluster.go [hunk 2]
```diff
@@ -345,9 +345,16 @@ func (ocb *openShiftClusterBackend) asyncOperationResultLog(log *logrus.Entry, i
...
+			log.Info("long running operation succeeded")
+			return
+		} else if err.StatusCode < 500 {
+			log = log.WithField("resultType", utillog.UserErrorResultType)
+		} else {
+			log = log.WithField("resultType", utillog.ServerErrorResultType)
+		}
 	} else {
 		log = log.WithField("resultType", utillog.ServerErrorResultType)
 	}
```
Comment (jaitaiwan) [inline pkg/backend/openshiftcluster.go:360]: This conditional could be simplified down.

**Review by jaitaiwan — APPROVED** [review]
LGTM small nit - the if statement seems unnecessarily complex

File: pkg/backend/openshiftcluster.go (hunk 2 above)
Comment (sankur-codes) [reply pkg/backend/openshiftcluster.go:360]: Based on the discussion above, the `if` statement might go away entirely once we discuss and check on the only instance of `CloudError` with 3XX `StatusCode`.

**Review by SrinivasAtmakuri — DISMISSED** [review]

**Comment by SrinivasAtmakuri** [comment]
Can we please add a few test cases coverage here? Thank you.

**Review by jaitaiwan — CHANGES_REQUESTED** [review]
A good point by Srini. Lets add some tests to go with this PR

**Comment by kartikaysharma27** [comment]
LGTM

**Review by mociarain — APPROVED** [review]
LGTM

**Review by SrinivasAtmakuri — APPROVED** [review]
LGTM.

**Review by jaitaiwan — APPROVED** [review]
LGTM

**Pull request merged**